### PR TITLE
MF-532 - Password change for authenticated user

### PR DESCRIPTION
**Pull request by mteodor**
### What does this do?
Adds an endpoint for changing password of currently authenticated user

### Which issue(s) does this PR fix/relate to?`
Put here `Resolves #XXX` to auto-close the issue that your PR fixes (if such)

### CODE REVIEW

File: docker/docker-compose.yml [hunk 1]
```diff
@@ -82,6 +82,16 @@ services:
       MF_USERS_GRPC_PORT: ${MF_USERS_GRPC_PORT}
       MF_USERS_SECRET: ${MF_USERS_SECRET}
       MF_JAEGER_URL: ${MF_JAEGER_URL}
+      MF_MAIL_DRIVER: ${MF_MAIL_DRIVER}
+      MF_MAIL_HOST: ${MF_MAIL_HOST}
+      MF_MAIL_PORT: ${MF_MAIL_PORT}
+      MF_MAIL_USERNAME: ${MF_MAIL_USERNAME}
+      MF_MAIL_PASSWORD: ${MF_MAIL_PASSWORD}
+      MF_MAIL_FROM_ADDRESS: ${MF_MAIL_FROM_ADDRESS}
+      MF_MAIL_FROM_NAME: ${MF_MAIL_FROM_NAME}
```
Comment (manuio) [inline docker/docker-compose.yml]: EMAIL

File: users/README.md [hunk 2]
```diff
@@ -34,6 +34,16 @@ default values.
...
 | MF_USERS_SECRET           | String used for signing tokens                                          | users          |
 | MF_JAEGER_URL             | Jaeger server URL                                                       | localhost:6831 |
+| MF_MAIL_DRIVER            | Mail server driver, mail server for sending reset password token        | smtp           |
+| MF_MAIL_HOST              | Mail server host                                                        | localhost      |
+| MF_MAIL_PORT              | Mail server port                                                        | 25             |
+| MF_MAIL_USERNAME          | Mail server username                                                    |                |
+| MF_MAIL_PASSWORD          | Mail server password                                                    |                |
+| MF_MAIL_FROM_ADDRESS      | Mail "from" address                                                     |                |
+| MF_MAIL_FROM_NAME         | Mail "from" name                                                        |                |
```
Comment (manuio) [inline users/README.md]: EMAIL

File: users/README.md [hunk 3]
```diff
@@ -66,6 +76,16 @@ services:
       MF_USERS_SERVER_CERT: [String path to server certificate in pem format]
       MF_USERS_SERVER_KEY: [String path to server key in pem format]
       MF_JAEGER_URL: [Jaeger server URL]
+      MF_MAIL_DRIVER: [Mail server driver smtp]
+      MF_MAIL_HOST: [MF_MAIL_HOST]
+      MF_MAIL_PORT: [MF_MAIL_PORT]
+      MF_MAIL_USERNAME: [MF_MAIL_USERNAME]
+      MF_MAIL_PASSWORD: [MF_MAIL_PASSWORD]
+      MF_MAIL_FROM_ADDRESS: [MF_MAIL_FROM_ADDRESS]
+      MF_MAIL_FROM_NAME: [MF_MAIL_FROM_NAME]
```
Comment (manuio) [inline users/README.md]: EMAIL

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/api/http/endpoint.go [hunk 4]
```diff
@@ -23,6 +24,55 @@ func registrationEndpoint(svc users.Service) endpoint.Endpoint {
...
+// email specified in the request.
+// Link contains token that has TTL that needs to be verified.
+// When user gets email with reset password link that hits this endpoint
+// which will return response into the ui form which then can
+// be used to enter email and new password and then request can be submitted
+// to the post endpoint which actually changes password.
+func passwordResetRequestEndpoint(svc users.Service) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		req := request.(passwResetReq)
+		res := resetPassRes{}
```
Comment (drasko) [inline users/api/http/endpoint.go]: Replace `Pass` with `Passwd`. It was uncler to me reading the code is the `pass` referencing to passing something...

File: users/api/http/responses.go [hunk 5]
```diff
@@ -46,3 +50,19 @@ func (res identityRes) Headers() map[string]string {
 func (res identityRes) Empty() bool {
 	return false
 }
+
+type resetPassRes struct {
```
Comment (drasko) [inline users/api/http/responses.go]: Replace `Pass` with `Passwd`

File: users/mocks/users.go [hunk 6]
```diff
@@ -47,3 +49,14 @@ func (urm *userRepositoryMock) RetrieveByID(ctx context.Context, email string) (
 
 	return val, nil
 }
+
+// UpdatePassword(
```
Comment (drasko) [inline users/mocks/users.go]: Please fix comment

File: users/service.go [hunk 7]
```diff
@@ -109,4 +141,37 @@ func (svc usersService) UserInfo(ctx context.Context, token string) (User, error
 		Password: "",
 		Metadata: dbUser.Metadata,
 	}, nil
+
+}
+
+func (svc usersService) GenerateResetToken(ctx context.Context, email, host string) error {
+
```
Comment (drasko) [inline users/service.go]: No need for white line here

File: users/token/token.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
+//Package token provides password recovery token generation with TTL
```
Comment (drasko) [inline users/token/token.go]: Put space after `//`

File: users/token/token.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+	envTokenSecret   = "MF_TOKEN_SECRET"
+	envTokenDuration = "MF_TOKEN_DURATION"
+	envTokenLogLevel = "MF_TOKEN_DEBUG_LEVEL"
+)
+
+var once sync.Once
+
+type tokenizer struct {
+	hmacSampleSecret []byte // secret for sigining token
+	tokenDuration    int    //token in duration in min
```
Comment (drasko) [inline users/token/token.go:46]: Space after `//`

File: users/token/token.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+	return instance().verify(tok)
+}
+
+// SendToken sends password recovery link to user
+func SendToken(host, email, token string) {
+	instance().sendToken(host, email, token)
+}
+
+func (t *tokenizer) generate(email string, offset int) (string, error) {
+
```
Comment (drasko) [inline users/token/token.go]: No need fro blank line

**Review by drasko — CHANGES_REQUESTED** [review]

**Pull request closed without merging**